item.lua

```diff
@@ -0,0 +1,340 @@
+    if x then -- item has a valid position
+        local igroup = dfhack.maps.getWalkableGroup(xyz2pos(x, y, z))
+        return not not wgroups[igroup]
+    else
+        return false
+    end
+end
+
+--- @return table<integer,boolean>
+function citizenWalkabilityGroups()
+    local cgroups = {}
+    for _, unit in pairs(dfhack.units.getCitizens(true)) do
+        local wgroup = dfhack.maps.getWalkableGroup(unit.pos)
+        cgroups[wgroup] = true
+    end
+    cgroups[0] = false -- exclude unwalkable tiles
+    return cgroups
+end
+
+function table.contains (t, val)
+    for _,v in ipairs(t) do
+        if v == val then return true end
+    end
+    return false
+end
+
+
+
+-----------------------------------------------------------------------
+-- external API: helpers to assemble filters and `execute` to execute.
+-----------------------------------------------------------------------
+
+--- @alias conditions (fun(item:item):boolean)[]
+
+--- @param tab conditions
+--- @param burrow burrow
+--- @param outside boolean
+function condition_burrow(tab,burrow, outside)
+    if outside then
+        table.insert(tab, function (item) return not containsItem(burrow, item) end)
+    else
+        table.insert(tab, function (item) return containsItem(burrow, item) end)
+    end
+end
+
+--- @param tab conditions
+--- @param match number|string
+function condition_type(tab, match)
+    if type(match) == "string" then
+        table.insert(
+            tab,
+            function (item) return df.item_type[item:getType()] == string.upper(match) end
+        )
+    elseif type(match) == "number" then
+        table.insert(
+            tab,
+            function (item) return item:getType() == type end
+        )
+    else qerror("match argument must be string or number")
+    end
+end
+
+--- @param tab conditions
+function condition_reachable(tab)
+    local cgroups = citizenWalkabilityGroups()
+    table.insert(tab, function(item) return fastReachable(item, cgroups) end)
+end
+
+--- @param tab conditions
+function condition_unreachable(tab)
+    local cgroups = citizenWalkabilityGroups()
+    table.insert(tab, function (item) return not fastReachable(item,cgroups) end)
+end
+
+-- uses the singular form without stack size (i.e., prickle berry)
+-- Exception: for corpse pieces like "wool", their description always includes stacksize (DF bug)
+--- @param tab conditions
+--- @param desc string
+function condition_description(tab, desc)
+    table.insert(
+        tab,
+        function(item) return dfhack.items.getDescription(item, 1) == desc end)
+end
+
+--- @param tab conditions
+--- @param material string
+function condition_material(tab, material)
+    table.insert(
+        tab,
+        function(item) return dfhack.matinfo.decode(item):toString() == material end)
+end
+
+--- @param tab conditions
+--- @param match string
+function condition_matcat(tab, match)
+    if table.contains(df.dfhack_material_category,match)
```

Comment:
why table.contains() instead of `if df.dfhack_material_category[match]`?

Reply:
Because `df.dfhack_material_category` maps numbers *to* material categories, and I didn't realize that the table is bidirectional. Given that 0, the value for "plant", is falsey, I guess the correct way is `if df.dfhack_material_category[match] ~= nil`

Reply:
the number `0` is truthy in Lua. the only falsey values are `nil` and `false`